Feature mediator coordination

nain-F49FF806 opened the pull request:
[Mediator Coordination Protocol](https://github.com/hyperledger/aries-rfcs/blob/main/features/0211-route-coordination/README.md)

- [x] MediateRequest
    - [x] Auth pubkey field (receive)
    - [x] Handling: Create account in storage database
    - [x] Respond with Grant if success
    - [x] Respond with Deny if already exist

- [x] MediateRequestDeny
    - [x] Reason field
- [ ] MediateRequestGrant
    - [x] No Field
    - [ ] Endpoint field (details undecided as of now, because serving on localhost)
    - [ ] Routing Keys field (unauthenticated for now, so not required)
  
- [x] Keylist-update
    - [x] Add
    - [x] Remove
- [x] Keylist-query
    - [x] Returns list of registered recipient keys
    - [ ] Pagination

nain-F49FF806 committed:
feat(sqlx-db): create migration 20230731173900_add_uniqueness_to_pubkey_feild.sql

nain-F49FF806 committed:
test(pickup): update e2e-pickup.rs with delivery-request tests

nain-F49FF806 committed:
feat(storage):  Rework MediatorPersistence for reworked DB tables

nain-F49FF806 committed:
feat(mediator-coordination): create coordination.rs

nain-F49FF806 committed:
feat(mediator-coordination): Enable Mediate request granting, along with database insertion requirements

Mediator Coordination Protocol
- [x] MediateRequest
    - [x] Auth pubkey field (receive)
    - [x] Handling: Create account in storage database
    - [x] Respond with Grant if success
    - [x] Respond with Deny if already exist
- [x] MediateRequestDeny
    - [x] Reason field
- [ ] MediateRequestGrant
    - [x] No Field
    - [ ] Endpoint field
    - [ ] Routing Keys field

nain-F49FF806 committed:
feature(forward): update forward.rs, patching minor clippy issue with absent error handling

Need to rework this code for better error handling

nain-F49FF806 committed:
feat(mediator-coordination): Support keylist-update message (add,remove recipient_key)

nain-F49FF806 committed:
feat(mediator-coordination): Support keylist-query

nain-F49FF806 committed:
test(pickup): Initialize receiver in database prior to pickup tests

nain-F49FF806 committed:
chore(lint):  Fix clippy warnings

nain-F49FF806 committed:
tes(forward)t: setup account before testing forward message persistence

Patrik-Stas commented on migrations/20230731173900_add_uniqueness_to_pubkey_feild.sql:
consider cleaning up migration folder, you know what the current desired state of the DB is, so you can create that in single migration

But it's good that you have tried the incremental migration approach, as people would tend to do it in production setting. But you have the luxury of not having to do so ;-)

nain-F49FF806 replied:
Yes :smile: . Sorry about the mess here. I was hoping to clean up this up, and start fresh with the finalised tables in a new repository. 
But I will do it right away here itself. :)

Patrik-Stas commented on src/routes/coordination.rs:
```diff
@@ -0,0 +1,170 @@
+    storage: Arc<T>,
+    mediate_req: MediateRequestData,
+) -> Json<MediatorCoordMsgEnum> {
+    let auth_pubkey = &mediate_req.auth_pubkey;
```
I suppose this `.auth_pubkey` is just temporary solution to differentiate these "accounts" which will later be done by the aries connection key?

nain-F49FF806 replied:
Yes. The signing key of the outer Aries connection will be used, when implemented. This extra field is used to distinguish clients until then.

Patrik-Stas commented on src/storage.rs:
```diff
@@ -24,53 +27,282 @@ pub async fn init() -> sqlx::MySqlPool {
+            .unwrap()
+            .get("account");
+        info!("Created account {:#?}, Adding auth_pubkey {:#?} as a default recipient_key to account", &account, &auth_pubkey);
+        sqlx::query("INSERT INTO recipients (account, recipient_key) VALUES (?, ?);")
```
The `auth_pubkey` represent the key for the counterparty having didcomm channel open with the mediator itself. This is not a recipient key a user would want to receive a message for. 

In this function you should only create the account, but not create any records in `recipients` table.

If I am missing something by a chance, please correct me.

nain-F49FF806 replied:
Initially (before implementing key-update), this was done to have `recipient_keys` pre-populated with something, for easier testing. It seemed like a sane default to use.

Now that key-update has been implemented, clients can explicitly add `recipient_keys`, so this has become vestigial. I shall remove it.

nain-F49FF806 replied:
Done

Patrik-Stas commented on src/storage.rs:
```diff
@@ -24,53 +27,282 @@ pub async fn init() -> sqlx::MySqlPool {
-        sqlx::query("INSERT INTO forward_raw VALUES (DEFAULT, ?, ?, DEFAULT)")
+    async fn create_account(&self, auth_pubkey: &str) -> Result<(), String> {
+        info!("Adding new account to database with auth_pubkey {:#?}", &auth_pubkey);
+        let insert_res = sqlx::query("INSERT INTO accounts (auth_pubkey) VALUES (?);")
```
Not part of this PR directly, but I see the schema for this table is currently this
```
CREATE TABLE IF NOT EXISTS accounts (
    id SERIAL,
    -- sequential time ordered uuid
    -- https://dev.mysql.com/blog-archive/mysql-8-0-uuid-support/
    account BINARY(16) DEFAULT (UUID_TO_BIN(UUID(),true)) NOT NULL UNIQUE,
    account_name VARCHAR(36) GENERATED ALWAYS AS (BIN_TO_UUID(account)) VIRTUAL,
    -- mediator facing pubkey
    auth_pubkey TEXT NOT NULL,
    -- live mode active or not
    live_delivery BOOLEAN NOT NULL DEFAULT FALSE,
    PRIMARY KEY(account)
);
```
Why not making `auth_pubkey` the primary key and get rid of both `account` and `id` cols? 

edit: actually I would remove also `auth_pubkey_sha` which was added in this PR's DB migration, why need that?

Patrik-Stas replied:
I'd consider removing `live_delivery` while we don't deal with it. I am not sure that's something that would reside in the permanent storage.
Unless you are sure, it's better to omit it while we don't use it.

nain-F49FF806 replied:
> Why not making `auth_pubkey` the primary key and get rid of both `account` and `id` cols?

This was done for 2 reasons:
- `auth_pubkey` being a "string", was not suitable to use as primary key or index. (MySQL complains)
  - The full size that a key can be isn't *explicitly* defined, so I've used STRING to store it, rather than fixed sized VARCHAR. This can be changed, as per your recommendation, in another comment. 
- To have the possibility to later change the `auth_pubkey` without having to create an entirely new 'account'. Basically to support let's say key rotation (manually, or through protocol). 
- Another later consideration is to have a primary identifier independent of user provided personal material. For ease in satisfying GDPR etc constraints. By having an independent identifier from the start we can support full deletion (including auth_key), while still keeping some relevant data, like say total forward requests served.

Because of 2,3 above, I think it'd still be good to have a separate 'account' column. `id` is superfluous and I can remove it.

nain-F49FF806 replied:
The `auth_pubkey_sha` can be useful to detect and deny duplicate registration for same `auth_key`. That's the only reason I added it. (Because iirc, STRING `auth_key` did not support UNIQUE constraint)

If switching to fixed size VARCHAR, I think this would not be needed.

nain-F49FF806 replied:
I have removed the `auth_pubkey_sha`, `live_delivery` columns. 
Changed type of `auth_pubkey` to VARCHAR since, we have expectation on size of keys. 
And thus am able to place UNIQUE constraint on `auth_pubkey` directly.

Patrik-Stas commented on src/storage.rs:
```diff
@@ -24,53 +27,282 @@ pub async fn init() -> sqlx::MySqlPool {
+    //     }
+    
+    // }
+    async fn persist_forward_message(&self, forward_msg: &ForwardMsg) -> Result<(), String> {
```
will be cleaner to pass `recipient_key: String, data: <String/binary type>`, the storage layer imo should really "know" about aries

nain-F49FF806 replied:
Thanks for the catch. Changing this.

nain-F49FF806 replied:
Done.

Patrik-Stas commented on src/storage.rs:
```diff
@@ -24,53 +27,282 @@ pub async fn init() -> sqlx::MySqlPool {
+    async fn persist_forward_message(&self, forward_msg: &ForwardMsg) -> Result<(), String> {
+        // Fetch recipients with given recipient_key
+        let mut rows = sqlx::query(
+            "SELECT * FROM recipients WHERE recipient_key = ?"
```
This would kill the service performance in practice, given your current schema

```
CREATE TABLE IF NOT EXISTS recipients (
    id SERIAL,
    account BINARY(16) NOT NULL,
    -- need not be unique (what if two accounts somehow register for same key?)
    -- the illegitimate won't be able to decrypt, even if they might be able to register reciept
    recipient_key TEXT NOT NULL,
    -- Consider using combination (recipient_key, account) as the primary key
    -- Done this way, using sha2 because recipient_key is TEXT (not directly indexable)
    -- recipient_key is text because key size, format may be variable
    recipient BINARY(32) GENERATED ALWAYS AS (
        UNHEX(
            SHA2(
                CONCAT((BIN_TO_UUID(account)), "+", recipient_key), 256
            )
        )
    ) STORED,
    PRIMARY KEY (recipient),
    FOREIGN KEY (account) REFERENCES accounts(account)
);
```
Currently you will end up scanning the entire table row-space.

The recipient key absolutely must be indexed and these values typically won't be huge. The public keys typically used in aries community are 32 bytes ed25519 keys (64bytes private key), which is at most 44 bs58 characters (we tend to use bs58 encoding). We should be able to get by with that, and we can alter the table if we need to increase in the future.

Patrik-Stas replied:
I also don't see the need for assuming multiple account would register same key. 
44 bytes = 352 bits, thats 9.173994e+105 

If someone is going to DDOS service and try to "register" 1 million of keys per second, it will still take them `3.6717431e+63` ( 2^(32*8)/(1000000 * 60 * 60 * 24 * 365)) years to exhaust the entire space, which is, well damn, less than atoms in universe, but it's a damn long time to be patient for :-D 

We can guard against malicious actors trying to take up key-space without actual possession of private key using different mechanisms.

You can make `recipient_key` primary 44 char indexed col and still sleep well at night ;-) Drop `recipient` along with that

nain-F49FF806 replied:
Had discussed this earlier with Bogdan (who initially had similar reservations :). 

The concern here is a possibility of a mischievous actor (say Bob) registering a _**known**_ valid recipient key (say of their conversation partner, Alice). Alice may or may not have pre-registered the key with the mediator before.  I.e. registration of known valid key. 

There would be 4 ways to handle this:

1. Deny registration of the existing recipient_key [_First come, first served_]
2. Allow registration, possibly replacing Alice's registration [_Latest is greatest_]
3. Allow registration, creating duplicate entry  [_Duplicate_]
4. Allow registration of key, scoped to Bob's account. [_Account scoped recipient_keys_] (_current_)

Consequences:

1. Is simplest, but can lead to 
  - User enumeration on mediator. Allowing a 3rd party (who obtains the recipient_key somehow) to find out if Alice has/had registered the key on a particular mediator. 
  - If Alice reuses a key for multiple parties (which Alice really shouldn't), this is especially more effective.
  - If Alice is reusing keys, then this could also let someone DoS Alice by registering the known key on other mediators where Alice hasn't yet registered. 
    - This may also be applicable if Alice is conversing with Bob and decides to use/switch mediator later. Bob can race with Alice in registering the known key.
2. Is an obvious DoS
3. May result in Bob being able to know if Alice has picked up some message. By asking for status of recipient key. Or ability to mark messages as read before Alice can fetch them. 
4. Allows Bob to subscribe to message forwarded to that recipient_key (albeit undecryptable). But avoids
    - DoS
    - knowledge if Alice has picked up the message. Because messages will be stored once for each recipient (account scoped recipient_key). 
    - Also avoids marking Alice's messages as read, since Alice, Bob will have different copies of the message.

Patrik-Stas replied:
Great analysis!

However as per our discussion oncall, we should go with option 1
> Deny registration of the existing recipient_key [First come, first served]

Which does have some corner case risks, but ultimately should be rather be solved on protocol level by requiring proof of key ownership upon key registration.
The issue described are fairly low likely corner cases - typically Alice will register her key in the mediator, before she shares it anywhere else. Alice also definitely should not reuse the same key for different connections. But it's valid corners, however, the protocol level solution should rather address that.

The other 3 options all lead to scenario where Bob can fetch a message address for Alice, be it encrypted, it's classified as personal data - under GDPR and also under other regulations/jurisdictions. So the other 3 options are simply no-go.

nain-F49FF806 replied:
Have changed the database accordingly. 
Commits are in the more recent (child) branch, [feature-pickup](https://github.com/nain-F49FF806/axum-test-server/pull/16)

nain-F49FF806 replied:
Thinking about a post-protocol update scenario, I realise there's another (not corner) case. :sweat_smile: 

Adding here for documentation, and to keep all the considerations in one place.

**Scenario**: The protocol includes proof of ownership for recipient_key :  a proof for example that Alice can decrypt (or sign) using recipient_key. 

- Now Alice may have multiple agents (mobile phone, laptop, cloud), each with unique Aries `auth_pubkey`. 
- Alice may want to register a `recipient_key` for each of her two or three agents, so Alice can get updates to whichever agent is handy. 
- Alice won't be able to do this, if we restrict `recipient_key` to one "account" (auth_pubkey) as per option 1 above. 

Now, this is a futuristic scenario, and we can perhaps think of it after the protocol work is done. 
But it does bring up an important effect of the restraint, even in our current scenario. 

Choosing 1 prevents Alice from registering owned recipient keys on multiple accounts of hers. 
So it would make subscribing to a recipient key on multiple devices more difficult. (Would have to sync `auth_pubkey` among agents to do so)

nain-F49FF806 replied:
_Suggested resolution:_ 
Until the protocol is updated, we should default to option1 to prevent leaking info to malicious parties that may subscribe without authorisation of recipient key owner.

Once authorisation/proof of ownership is baked into the protocol, we should reconsider.

nain-F49FF806 replied:
Note: Syncing of recipient keys would have to be done anyway. So unless `auth_pubkey` is hardware backed and tied to the device, it could possibly also be synced, obviating the need for tying multiple `auth_pubkey` keys to one account.

Patrik-Stas commented on src/storage.rs:
```diff
@@ -24,53 +27,282 @@ pub async fn init() -> sqlx::MySqlPool {
+        }
+        Ok(())
     }
     async fn retrieve_pending_message_count(&self, recipient_key: Option<&String>) -> u32 {
```
this function too should have `auth_pubkey` argument - counting messages must be limited to the keys owned by a particular account

Patrik-Stas replied:
if you follow my advice about `recipients` above, this function will as well greatly simplify

nain-F49FF806 replied:
> this function too should have `auth_pubkey` argument - counting messages must be limited to the keys owned by a particular account

Done.

nain-F49FF806 replied:
I have simplified the function, please have a look.  

https://github.com/nain-F49FF806/axum-test-server/blob/9df1ef5bfe7caf9d96b26c0f0790ca2b0048776b/src/storage.rs#L136

Turns out, it was unnecessarily complex regardless of the recipient situation (which I will tackle next). Using the `auth_pubkey` argument adds sufficient context to simply it greatly.

Patrik-Stas reviewed (changes requested):
left a review, my comments are mostly rooting out from the current db schema design. As they say: "keep it simple stupid" applies here imo :-)

nain-F49FF806 committed:
Remove dependency on custom structures from trait method.

Pass recipient_key, message_data directly instead

nain-F49FF806 committed:
feature(pickup): Limit status read to authorized key holder

nain-F49FF806 committed:
test(pickup, mediator-coord): Add auth_pubkey to status requests

nain-F49FF806 committed:
feature(storage): SImplify account fetching using new function

nain-F49FF806 committed:
feature(pickup, storage): Simplify retrieve_pending_message_count

Also fix glaring bug.
The existing implementation was for previous table structure and test case, and modified improperly in place.

Now it is correctly ported to new table structure.

nain-F49FF806 committed:
feature(storage, db):  Simplify MySQL tables and focus on primary key index efficiency

nain-F49FF806 commented on migrations/20230810132416_3-part-data-(accounts,_recipients,_messages).sql:
Have cleaned up the tables. Now they are much leaner.

Also, indexed (using primary key) on relevant fields, which will be searched.

nain-F49FF806 committed:
test(forward): Deliver to recipient key, not mediator facing pubkey

nain-F49FF806 committed:
fix(mediator-coord): Do not add auth_pubkey as recipient_key.

Client can do so explicitly using key-update request

nain-F49FF806 committed:
feature(storage): Fix little bug, swapped variables.

Patrik-Stas commented on migrations/20230810132416_3-part-data-(accounts,_recipients,_messages).sql:
```diff
@@ -0,0 +1,37 @@
+    seq_num SERIAL,
+    -- sequential time ordered uuid
+    -- https://dev.mysql.com/blog-archive/mysql-8-0-uuid-support/
+    account BINARY(16) DEFAULT (UUID_TO_BIN(UUID(),true)) NOT NULL UNIQUE,
```
- rename to `account_id` or just `id`

nain-F49FF806 replied:
Done

Patrik-Stas commented on migrations/20230810132416_3-part-data-(accounts,_recipients,_messages).sql:
```diff
@@ -0,0 +1,37 @@
+    -- for display purpose
+    account_name CHAR(36) GENERATED ALWAYS AS (BIN_TO_UUID(account)) VIRTUAL,
+    -- mediator facing pubkey
+    auth_pubkey VARCHAR(64) NOT NULL UNIQUE,
```
this should be indexed

nain-F49FF806 replied:
It seems UNIQUE keys are automatically indexed. (Found below when searching for info to learn about indexing)
[ref1](https://stackoverflow.com/questions/366186/is-the-sql-server-unique-key-also-an-index)
[ref2](https://dba.stackexchange.com/a/2855)

Have added comment mentioning the same. And ensured indexing for other places (`recipient_keys`)

Patrik-Stas commented on migrations/20230810132416_3-part-data-(accounts,_recipients,_messages).sql:
```diff
@@ -0,0 +1,37 @@
+
+CREATE TABLE IF NOT EXISTS recipients (
+    seq_num SERIAL,
+    account BINARY(16) NOT NULL,
```
rename
